### PR TITLE
common: implementing finalize_bundle.

### LINKED ISSUE #613: Model/bundle: About bundle finalization
Now we have `calculate_bundle_hash` to get bundle hash, 
I think a `bundle_finalize` function to fill in bundle hash through transactions is also needed.

Any idea?

### PR DESCRIPTION
**Pull request by oopsmonk**
implementing finalize_bundle and the test case. 

Fixes #613 

# Test Plan:
CI must pass

### CODE REVIEW

File: common/model/bundle.c [hunk 1]
```diff
@@ -51,3 +51,55 @@ void calculate_bundle_hash(bundle_transactions_t *bundle, flex_trit_t *out) {
...
+  byte_t normalized_hash[HASH_LENGTH_TRYTE];
+  trit_t increased_tag_trits[NUM_TRITS_TAG];
+  flex_trit_t bundle_hash[FLEX_TRIT_SIZE_243];
+  trit_t one[NUM_TRITS_TAG] = {
+      1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+
+  Kerl kerl = {};
```
Comment (th0br0) [inline common/model/bundle.c]: Let's have the user pass in a Kerl instance.

File: common/model/bundle.c [hunk 2]
```diff
@@ -51,3 +51,55 @@ void calculate_bundle_hash(bundle_transactions_t *bundle, flex_trit_t *out) {
...
+                              NUM_TRITS_TAG);
+        goto update_hash;
+      }
+    }
+    valid_bundle = true;
+  }
+}
+
+// for debugging
+void dump_bundle(bundle_transactions_t *bundle) {
```
Comment (th0br0) [inline common/model/bundle.c]: guard this with `#ifdef DEBUG`?

File: common/model/bundle.c [hunk 3]
```diff
@@ -51,3 +51,55 @@ void calculate_bundle_hash(bundle_transactions_t *bundle, flex_trit_t *out) {
   flex_trits_from_trits(out, NUM_TRITS_HASH, bundle_hash_trits, NUM_TRITS_HASH,
                         NUM_TRITS_HASH);
 }
+
+void finalize_bundle(bundle_transactions_t *bundle) {
```
Comment (th0br0) [inline common/model/bundle.c]: could we call this `bundle_finalize`?

File: common/model/bundle.c (hunk 1 above)
Comment (th0br0) [reply common/model/bundle.c]: this also goes for `calculate_bundle_hash` - less memory :)

File: common/model/bundle.c [hunk 4]
```diff
@@ -51,3 +51,55 @@ void calculate_bundle_hash(bundle_transactions_t *bundle, flex_trit_t *out) {
...
+    // normalize
+    normalize_hash(bundle_hash, normalized_hash);
+    kerl_reset(&kerl);
+    // checking 'M'
+    for (int i = 0; i < HASH_LENGTH_TRYTE; i++) {
+      if (normalized_hash[i] == 13) {
+        // Insecure bundle. Increment Tag and recompute bundle hash.
+        head_tx = (iota_transaction_t *)utarray_front(bundle);
+        flex_trits_to_trits(increased_tag_trits, NUM_TRITS_TAG,
+                            transaction_obsolete_tag(head_tx), NUM_TRITS_TAG,
```
Comment (th0br0) [inline common/model/bundle.c]: use `transaction_set_obsolete_tag`

File: common/model/bundle.c [hunk 5]
```diff
@@ -51,3 +51,55 @@ void calculate_bundle_hash(bundle_transactions_t *bundle, flex_trit_t *out) {
...
+                            NUM_TRITS_TAG);
+        add_trits(one, increased_tag_trits, NUM_TRITS_TAG);
+        flex_trits_from_trits(transaction_obsolete_tag(head_tx), NUM_TRITS_TAG,
+                              increased_tag_trits, NUM_TRITS_TAG,
```
Comment (th0br0) [inline common/model/bundle.c:77]: use `transaction_set_obsolete_tag`

File: common/model/bundle.c (hunk 4 above)
Comment (th0br0) [reply common/model/bundle.c]: also, you just need to get this tag *once* and then increment by one each time. no need to read every time.

File: common/model/bundle.c [hunk 6]
```diff
@@ -51,3 +51,55 @@ void calculate_bundle_hash(bundle_transactions_t *bundle, flex_trit_t *out) {
...
+    kerl_reset(&kerl);
+    // checking 'M'
+    for (int i = 0; i < HASH_LENGTH_TRYTE; i++) {
+      if (normalized_hash[i] == 13) {
+        // Insecure bundle. Increment Tag and recompute bundle hash.
+        head_tx = (iota_transaction_t *)utarray_front(bundle);
+        flex_trits_to_trits(increased_tag_trits, NUM_TRITS_TAG,
+                            transaction_obsolete_tag(head_tx), NUM_TRITS_TAG,
+                            NUM_TRITS_TAG);
+        add_trits(one, increased_tag_trits, NUM_TRITS_TAG);
```
Comment (th0br0) [inline common/model/bundle.c]: could you make this a const / static outside of the function scope?

File: common/model/transaction.c [hunk 7]
```diff
@@ -347,3 +347,65 @@ size_t transaction_deserialize_from_trits(iota_transaction_t *const transaction,
 void transaction_free(iota_transaction_t *const transaction) {
   free(transaction);
 }
+
+// for debugging
+void transaction_obj_dump(iota_transaction_t *tx_obj) {
```
Comment (th0br0) [inline common/model/transaction.c]: guard with `#ifdef DEBUG`

File: common/model/transaction.h [hunk 8]
```diff
@@ -13,6 +13,8 @@ extern "C" {
 #define __COMMON_MODEL_TRANSACTION_H_
 
 #include <assert.h>
+#include <inttypes.h>
```
Comment (th0br0) [inline common/model/transaction.h]: not needed?

File: common/model/transaction.h [hunk 9]
```diff
@@ -13,6 +13,8 @@ extern "C" {
...
+#include <stdio.h>
```
Comment (th0br0) [inline common/model/transaction.h]: not needed

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: common/model/transaction.h (hunk 9 above)
Comment (oopsmonk) [reply common/model/transaction.h]: for debugging.

File: common/model/transaction.c (hunk 7 above)
Comment (oopsmonk) [reply common/model/transaction.c]: sure.

File: common/model/bundle.c (hunk 5 above)
Comment (oopsmonk) [reply common/model/bundle.c:77]: it's inefficient in this case.

File: common/model/bundle.c (hunk 4 above)
Comment (oopsmonk) [reply common/model/bundle.c]: need to update the `obsolete_tag` for `calculate_bundle_hash`, isn't it?

File: common/model/bundle.c (hunk 3 above)
Comment (oopsmonk) [reply common/model/bundle.c]: sure, or `bundle_finalization`? b/c finalize is a verb.

File: common/model/bundle.c (hunk 1 above)
Comment (oopsmonk) [reply common/model/bundle.c]: yup, it's a quick approach to ease `prepare_transfers` implementation. I think except the Kerl, I am trying to reduce flex_trits conversions here, later. so an optimization version comes after the APIs. what do you think?

File: common/model/bundle.c (hunk 3 above)
Comment (oopsmonk) [reply common/model/bundle.c]: btw, `finalize_bundle` is the similar naming rule with `calculate_bundle_hash`.

File: common/model/bundle.c [hunk 10]
```diff
@@ -51,3 +51,55 @@ void calculate_bundle_hash(bundle_transactions_t *bundle, flex_trit_t *out) {
...
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+
+  Kerl kerl = {};
+  init_kerl(&kerl);
+
+  while (!valid_bundle) {
+  update_hash:
+    // bundel hash
```
Comment (thibault-martinez) [inline common/model/bundle.c]: Typo

File: common/model/bundle.c [hunk 11]
```diff
@@ -51,3 +51,55 @@ void calculate_bundle_hash(bundle_transactions_t *bundle, flex_trit_t *out) {
...
+    calculate_bundle_hash(bundle, bundle_hash);
+    // update bundle_hash
+    BUNDLE_FOREACH(bundle, curr_tx) {
+      transaction_set_bundle(curr_tx, bundle_hash);
+    }
```
Comment (thibault-martinez) [inline common/model/bundle.c]: Set the bundle hash to txs only when it's valid ?

File: common/model/bundle.c [hunk 12]
```diff
@@ -51,3 +51,55 @@ void calculate_bundle_hash(bundle_transactions_t *bundle, flex_trit_t *out) {
...
+        // Insecure bundle. Increment Tag and recompute bundle hash.
+        head_tx = (iota_transaction_t *)utarray_front(bundle);
+        flex_trits_to_trits(increased_tag_trits, NUM_TRITS_TAG,
+                            transaction_obsolete_tag(head_tx), NUM_TRITS_TAG,
+                            NUM_TRITS_TAG);
+        add_trits(one, increased_tag_trits, NUM_TRITS_TAG);
+        flex_trits_from_trits(transaction_obsolete_tag(head_tx), NUM_TRITS_TAG,
+                              increased_tag_trits, NUM_TRITS_TAG,
+                              NUM_TRITS_TAG);
+        goto update_hash;
```
Comment (thibault-martinez) [inline common/model/bundle.c:79]: Just use `continue;` ?

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: common/model/bundle.c (hunk 11 above)
Comment (oopsmonk) [reply common/model/bundle.c]: right, thanks :)

File: common/model/bundle.c (hunk 12 above)
Comment (oopsmonk) [reply common/model/bundle.c:79]: nice try :-P

File: common/model/bundle.c (hunk 2 above)
Comment (oopsmonk) [reply common/model/bundle.c]: done.

File: common/model/transfer.h [hunk 13]
```diff
@@ -20,7 +20,10 @@ extern "C" {
 #include "common/trinary/trit_long.h"
 #include "utils/logger_helper.h"
 
-#define NUM_TRITS_ESSENCE 486
+// bundle essence = 486 trits
+#define NUM_TRITS_ESSENCE                                         \
+  (NUM_TRITS_ADDRESS + NUM_TRITS_VALUE + NUM_TRITS_OBSOLETE_TAG + \
+   NUM_TRITS_TIMESTAMP + NUM_TRITS_CURRENT_INDEX + NUM_TRITS_LAST_INDEX)
```
Comment (thibault-martinez) [inline common/model/transfer.h]: Can you put this in `transaction.h` with the other sizes ?

File: common/model/tests/test_bundle.c [hunk 14]
```diff
@@ -33,10 +33,78 @@ void test_normalized_bundle(void) {
...
+  transaction_set_obsolete_tag(&tx, tag);
+  transaction_set_timestamp(&tx, 1546760811);
+  transaction_set_value(&tx, 0);
+  transaction_set_current_index(&tx, 2);
+  transaction_set_last_index(&tx, 2);
+  tx.loaded_columns_mask |= MASK_ALL_COLUMNS;
+  bundle_transactions_add(bundle, &tx);
+
+  bundle_finalize(bundle);
+  bundle_dump(bundle);
```
Comment (thibault-martinez) [inline common/model/tests/test_bundle.c:97]: Remove

File: common/model/bundle.c (hunk 4 above)
Comment (th0br0) [reply common/model/bundle.c]: yes. but you don't need to `flex_trits_to_trits` it each time, only once. (and then increment that one)

File: common/model/bundle.c (hunk 5 above)
Comment (th0br0) [reply common/model/bundle.c:77]: oh you're right, this is trits->flex_trits in place.

File: common/model/transaction.h (hunk 9 above)
Comment (th0br0) [reply common/model/transaction.h]: you can include these in the .c though? no need to depend on them for all includers of this header

File: common/model/bundle.c (hunk 1 above)
Comment (th0br0) [reply common/model/bundle.c]: it's a minimal change, at least file an issue?

File: common/model/bundle.c [hunk 15]
```diff
@@ -51,3 +55,52 @@ void calculate_bundle_hash(bundle_transactions_t *bundle, flex_trit_t *out) {
...
+    kerl_reset(&kerl);
+    // checking 'M'
+    for (int i = 0; i < HASH_LENGTH_TRYTE; i++) {
+      if (normalized_hash[i] == 13) {
+        // Insecure bundle. Increment Tag and recompute bundle hash.
+        head_tx = (iota_transaction_t *)utarray_front(bundle);
+        flex_trits_to_trits(increased_tag_trits, NUM_TRITS_TAG,
+                            transaction_obsolete_tag(head_tx), NUM_TRITS_TAG,
+                            NUM_TRITS_TAG);
+        add_trits(one, increased_tag_trits, NUM_TRITS_TAG);
```
Comment (thibault-martinez) [inline common/model/bundle.c]: I think you can do `add_assign(increased_tag_trits, NUM_TRITS_TAG, 1)`; hence no need for that `one` static array

**Review by th0br0 — CHANGES_REQUESTED** [review]

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: common/model/tests/test_bundle.c (hunk 14 above)
Comment (oopsmonk) [reply common/model/tests/test_bundle.c:97]: why?

Comment (thibault-martinez) [reply common/model/tests/test_bundle.c:97]: Assertions are sufficient, IMHO we shouldn't pollute stdout with logs if we are looking for information...

File: common/model/bundle.c (hunk 15 above)
Comment (oopsmonk) [reply common/model/bundle.c]: cool, confused about the name of `add_assign`, could you explain more?

Comment (thibault-martinez) [reply common/model/bundle.c]: It modifies the trits you give it by operating a trit addition of an index on it.
For example, it's used to obtain a subseed by adding an index to a seed: https://github.com/iotaledger/entangled/blob/develop/common/sign/v1/iss.c.inc#L22

File: common/model/bundle.c (hunk 1 above)
Comment (oopsmonk) [reply common/model/bundle.c]: sure, filed in #613 
just don't want to change consensus/bundle_validator atm and keep this PR in the scope of the bundle model. :-)

File: common/model/tests/test_bundle.c (hunk 14 above)
Comment (oopsmonk) [reply common/model/tests/test_bundle.c:97]: IMO, it only shows up when `DEBUG` flag is applied and additional info are useful in the test case.

File: common/model/bundle.c (hunk 4 above)
Comment (oopsmonk) [reply common/model/bundle.c]: you're right, thanks.

File: common/model/tests/test_bundle.c (hunk 14 above)
Comment (thibault-martinez) [reply common/model/tests/test_bundle.c:97]: Whoops I overlooked, you're right

**Review by thibault-martinez — APPROVED** [review]

File: common/model/bundle.c (hunk 1 above)
Comment (oopsmonk) [reply common/model/bundle.c]: done.

File: common/model/bundle.c [hunk 16]
```diff
@@ -29,25 +29,71 @@ void bundle_transactions_add(bundle_transactions_t *const bundle,
...
+
+void bundle_finalize(bundle_transactions_t *bundle, Kerl *const kerl) {
+  iota_transaction_t *curr_tx = NULL;
+  bool valid_bundle = false;
+  iota_transaction_t *head_tx = NULL;
+  byte_t normalized_hash[HASH_LENGTH_TRYTE];
+  trit_t increased_tag_trits[NUM_TRITS_TAG];
+  flex_trit_t bundle_hash[FLEX_TRIT_SIZE_243];
+
+  init_kerl(kerl);
```
Comment (th0br0) [inline common/model/bundle.c]: not needed. reset inside the loop via `bundle_calculate_hash`

File: common/model/bundle.c [hunk 17]
```diff
@@ -29,25 +29,71 @@ void bundle_transactions_add(bundle_transactions_t *const bundle,
...
+
+  head_tx = (iota_transaction_t *)utarray_front(bundle);
+  flex_trits_to_trits(increased_tag_trits, NUM_TRITS_TAG,
+                      transaction_obsolete_tag(head_tx), NUM_TRITS_TAG,
+                      NUM_TRITS_TAG);
+  while (!valid_bundle) {
+  update_hash:
+    // bundle hash
```
Comment (th0br0) [inline common/model/bundle.c:67]: `kerl_reset(kerl)` is missing here

File: common/model/bundle.c [hunk 18]
```diff
@@ -29,25 +29,71 @@ void bundle_transactions_add(bundle_transactions_t *const bundle,
...
-void calculate_bundle_hash(bundle_transactions_t *bundle, flex_trit_t *out) {
+void bundle_calculate_hash(bundle_transactions_t *bundle, Kerl *const kerl,
+                           flex_trit_t *out) {
   iota_transaction_t *curr_tx = NULL;
   trit_t essence_trits[NUM_TRITS_ESSENCE];
-  Kerl kerl = {};
-  init_kerl(&kerl);
-
   trit_t bundle_hash_trits[NUM_TRITS_HASH];
+  init_kerl(kerl);
```
Comment (th0br0) [inline common/model/bundle.c:37]: kerl_reset(kerl)

File: common/model/tests/test_bundle.c [hunk 19]
```diff
@@ -33,10 +33,80 @@ void test_normalized_bundle(void) {
...
 
+void test_bundle_finalize(void) {
+  bundle_transactions_t *bundle = NULL;
+  flex_trit_t addr_sender[FLEX_TRIT_SIZE_243] = {};
+  flex_trit_t addr_receiver[FLEX_TRIT_SIZE_243] = {};
+  flex_trit_t tag[FLEX_TRIT_SIZE_81] = {};
+  flex_trit_t obsolete_tag[FLEX_TRIT_SIZE_81] = {};
+  iota_transaction_t tx;
+  iota_transaction_t *bundle_tx;
+  Kerl kerl = {};
```
Comment (th0br0) [inline common/model/tests/test_bundle.c:44]: call `init_kerl` on this

File: common/model/transaction.c [hunk 20]
```diff
@@ -347,3 +351,67 @@ size_t transaction_deserialize_from_trits(iota_transaction_t *const transaction,
...
```
Comment (th0br0) [inline common/model/transaction.c]: could we `#ifdef DEBUG` the whole method not just the impl/

File: common/model/transaction.h [hunk 21]
```diff
@@ -399,6 +404,8 @@ static inline void transaction_set_arrival_timestamp(
  ***********************************************************************************************************/
 
 void transaction_reset(iota_transaction_t *const transaction);
+// for debugging
+void transaction_obj_dump(iota_transaction_t *tx_obj);
```
Comment (th0br0) [inline common/model/transaction.h:408]: ifdef debug the method here as well?

File: consensus/bundle_validator/bundle_validator.c [hunk 22]
```diff
@@ -135,6 +135,7 @@ retcode_t iota_consensus_bundle_validator_validate(
   int64_t bundle_value = 0;
   flex_trit_t bundle_hash[FLEX_TRIT_SIZE_243];
   bool valid_sig = true;
+  Kerl kerl = {};
```
Comment (th0br0) [inline consensus/bundle_validator/bundle_validator.c:141]: call `init_kerl` on this?

**Review by th0br0 — CHANGES_REQUESTED** [review]

**Comment by th0br0** [comment]
almost there! :)

File: common/model/bundle.c (hunk 18 above)
Comment (oopsmonk) [reply common/model/bundle.c:37]: init_kerl == kerl_reset, do we actully need `kerl_reset`?

File: common/model/bundle.c (hunk 17 above)
Comment (oopsmonk) [reply common/model/bundle.c:67]: at line#73, but it's redundant, let's just make sure kerl has initialized before it's called.

File: common/model/tests/test_bundle.c (hunk 19 above)
Comment (oopsmonk) [reply common/model/tests/test_bundle.c:44]: it's called inside the bundle method.

File: consensus/bundle_validator/bundle_validator.c (hunk 22 above)
Comment (oopsmonk) [reply consensus/bundle_validator/bundle_validator.c:141]: kerl has initialized in the bundle method.

File: common/model/bundle.c (hunk 18 above)
Comment (th0br0) [reply common/model/bundle.c:37]: no, my bad. thought there was a difference!

**Review by th0br0 — APPROVED** [review]

**Pull request merged**